statsmodels/sandbox/mice/mice.py

```diff
@@ -0,0 +1,1287 @@
+            imp.data.to_csv('data%02d.csv' % j)
+            j += 1
+"""
+
+
+class PatsyFormula(object):
+    """
+    A simple wrapper for a string to be interpreted as a Patsy formula.
+    """
+    def __init__(self, formula):
+        self.formula = "0 + " + formula
+
+
+class MICEData(object):
+
+    __doc__ = """\
+    Wrap a data set to allow missing data handling with MICE.
+
+    Parameters
+    ----------
+    data : Pandas data frame
+        The data set.
+    perturbation_method : string
+        The default perturbation method
+    k_pmm : int
+        The number of nearest neighbors to use during predictive mean
+        matching.  Can also be specified in `fit`.
+    history_callback : function
+        A function that is called after each complete imputation
+        cycle.  The return value is appended to `history`.  The
+        MICEData object is passed as the sole argument to
+        `history_callback`.
+
+    Examples
+    --------
+    Draw 20 imputations from a data set called `data` and save them in
+    separate files with filename pattern `dataXX.csv`.  The variables
+    other than `x1` are imputed using linear models fit with OLS, with
+    mean structures containing main effects of all other variables in
+    `data`.  The variable named `x1` has a condtional mean structure
+    that includes an additional term for x2^2.
+    %(_mice_data_example_1)s
+
+    Impute using default models, using the MICEData object as an
+    iterator.
+    %(_mice_data_example_2)s
+    Notes
+    -----
+    Allowed perturbation methods are 'gaussian' (the model parameters
+    are set to a draw from the Gaussian approximation to the posterior
+    distribution), and 'boot' (the model parameters are set to the
+    estimated values obtained when fitting a bootstrapped version of
+    the data set).
+
+    `history_callback` can be implemented to have side effects such as
+    saving the current imputed data set to disk.
+    """ % {'_mice_data_example_1': _mice_data_example_1,
+           '_mice_data_example_2': _mice_data_example_2}
+
+    def __init__(self, data, perturbation_method='gaussian',
+                 k_pmm=20, history_callback=None):
+
+        # Drop observations where all variables are missing.  This
+        # also has the effect of copying the data frame.
+        self.data = data.dropna(how='all').reset_index(drop=True)
+
+        self.history_callback = history_callback
+        self.history = []
+        self.predict_kwds = {}
+
+        # Assign the same perturbation method for all variables.
+        # Can be overriden when calling 'set_imputer'.
+        self.perturbation_method = defaultdict(lambda :
+                                               perturbation_method)
+
+        # Map from variable name to indices of observed/missing
+        # values.
+        self.ix_obs = {}
+        self.ix_miss = {}
+        for col in self.data.columns:
+            ix_obs, ix_miss = self._split_indices(self.data[col])
```

Comment:
I usually prefer unique names for different things to avoid getting confused
e.g ix_obs_coll, ix_miss_coll   (or often I add an `_i` for one element